kernel/declarations.ml

```diff
@@ -102,24 +102,75 @@ type typing_flags = {
+    so, an abstr_info is both the context x:T,z:V(x) to generalize
+    (skipping y which does not occur), and the substitution [a↦x,c↦z]
+    where in practice, x and z are canonical (hence implicit) de
+    Bruijn indices, that is, only the instantiation [a,c] is kept *)
 
-(** Data needed to abstract over the section variable and universe hypotheses *)
 type abstr_info = {
-  abstr_ctx : Constr.named_context;
-  (** Section variables of this prefix *)
-  abstr_subst : Univ.Instance.t;
-  (** Actual names of the abstracted variables *)
+  abstr_ctx : Constr.rel_context;
+  (** Context over which to generalize (e.g. x:T,z:V(x)) *)
   abstr_uctx : Univ.AbstractContext.t;
-  (** Universe quantification, same length as the substitution *)
+  (** Universe context over which to generalize *)
+  abstr_subst : Id.t list;
+  (** Canonical renaming represented by its domain made of the
+      actual names of the abstracted term variables (e.g. [a,c]);
+      the codomain made of de Bruijn indices is implicit *)
+  abstr_usubst : Univ.universe_level_subst;
+  (** Universe substitution *)
+}
+
+(** The instantiation to apply to generalized declarations so that
+    they behave as if not generalized: this is the a1..an instance to
+    apply to a declaration c in the following transformation:
+    [a1:T1..an:Tn, C:U(a1..an) ⊢ v(a1..an,C):V(a1..an,C)
+     ~~>
+     C:Πx1..xn.U(x1..xn), a1:T1..an:Tn ⊢ v(a1..an,Ca1..an):V(a1..an,Ca1..an)]
+    note that the data looks close to the one for substitution above
+    (because the substitution are represented by their domain) but
+    here, local definitions of the context have been dropped *)
+
+type abstr_inst_info = {
+  abstr_inst : Constr.t array;
+  (** The variables to reapply (excluding "let"s of the context) *)
+  abstr_uinst : Univ.Instance.t;
+  (** Abstracted universe variables to reapply *)
 }
 
+(** The collection of instantiations to apply to generalized
+    declarations so that they behave as if not generalized.
+    This accounts for the permutation (lambda-lifting) of global and
+    local declarations.
+    Using the notations above, a expand_info is a map [c ↦ a1..an]
+    over all generalized global declarations of the section *)
+
+type 'a entry_map = 'a Cmap.t * 'a Mindmap.t
+type expand_info = abstr_inst_info entry_map
+
+(** The collection of instantiations to be done on generalized
+    declarations + the generalization to be done on a specific
+    judgment:
+    [a1:T1,a2:T2,C:U(a1) ⊢ v(a1,a2,C):V(a1,a2,C)
+     ~~>
+     c:Πx.U(x) ⊢ λx1x2.(v(a1,a2,cx1)[a1,a2:=x1,x2]):Πx1x2.(V(a1,a2,ca1)[a1,a2:=x1,x2])]
+    so, a cooking_info is the map [c ↦ x1..xn],
+    the context x:T,y:U to generalize, and the substitution [x,y] *)
+
 type cooking_info = {
-  modlist : work_list;
-  abstract : abstr_info;
+  expand_info : expand_info;
+  abstr_info : abstr_info;
+  abstr_inst_info : abstr_inst_info; (* relevant for recursive types *)
+  names_info : Id.Set.t; (* set of generalized names *)
 }
```

Comment:
I think that there is too much stuff exposed in this API. The datatypes used to live in `Cooking` and were moved here by @SkySkimmer IIRC but now that we're supposed to preserve more invariants I believe this should be move back under an opaque interface. How doable is this?

Reply:
It was in opaqueproof, not cooking
b1675a61ac5bee623995df785b087fa785885dad

Reply:
Right, right. But now that `Opaqueproof` is parameterized by this data, it should be easier to put it in `Cooking`, shouldn't it?

Reply:
Dependency order goes `opaqueproof <- declarations <- cooking` so I guess no

Reply:
This should still be tried nonetheless, I really don't like having all these internal bits exposed to everybody.

Reply:
A possibility would be to split `cooking.ml` into two parts:
- the low-level infrastructure up to `subst_private_univs` (and with an abstract export of `cooking_info`), coming before `opaqueproof.ml`
- the high-level commands `cook_constant` and `cook_inductive` coming after `declarations` (with possible integration to `term_typing.ml` if we don't want creating an extra file)

At worst, in the higher levels (vernac, tactics, ...) only the field `abstr_inst` really needs to be visible (and it could as well be hidden behind an access function).

Reply:
I pushed a commit in this directon, providing a better abstraction of `cooking_info`. We could abstract even a bit further, depending on feedback.